### ISSUE TITLE
[nearly-closing] Invoke-MsBuild does not exit?

### ISSUE THREAD

**Issue by StefanoBettega** [deadlydog/Invoke-MsBuild#11]
Hello, we are using Invoke-MsBuild to automate projects compilation, so that we can build libraries and then a native NuGet package all from command line (maybe on a build machine, later).

We call Invoke-MsBuild from a powershell script that cycles on configurations, platform, toolset we need (all of them read from an XML configuration file that is per-project).

Testing all of this on several machines gives sometimes strange behaviors: in fact it seems that Invoke-MsBuild never ends, and the invoker powershell script hangs waiting an output that never comes.

This seems to happen especially when changing from a build configuration to a different configuration or platform (i.e. from Debug to Release, or from Win32 to x64).

Is anybody experiencing such problem? I also tried to use the **/p:UseSharedConfiguration=false** workaround suggested [on your blog](http://blog.danskingdom.com/tag/msbuild-msbuild-exe-slow-hangs-build-compile/) but it seems not to help.

Thanks in advance,
Stefano

**Update 1**
I left my build machine alone awaiting for the command to terminate and I can confirm that after a delay of about 10 minutes, the build process succesfully exits. I double checked and the UseSharedConfiguration workaround is correctly applied, but it doesn't work.

**Comment by StefanoBettega** [comment]
Ok, aftert googling around a little bit more, I found [this topic](http://stackoverflow.com/questions/19905639/powershell-start-process-with-wait-option-taking-long-time-to-return) on stackoverflow that seems to solve the question.

Adding **/nodeReuse:false** to MSBuild command line seems to work, and build time is now acceptable.
Will wait more test then close this issue. You might want to integrate this workaround in Invoke-MSBuild as additional switch in case of issues, like Use32BitMsBuild.

**Comment by timker** [comment]
I also get this!, especially on  the first build in my CI server

I've also tried /nr:false this to not avail:
http://nftb.saturdaymp.com/fixing-msbuild-not-exiting/

**Comment by StefanoBettega** [comment]
**tl;dr**
Please find attached my modified version of Invoke-MsBuild.psm1 powershell module. You have to invoke it in -PassThru mode but it works as in Wait mode, and it seems to work correctly.

Let me know your opinions.

----
**The possible solution**
Ok, I thought I found a solution yesterday, so I changed the title to [nearly closing] and I posted a comment. But after some tests i found it didn't work, so I reverted changes to title and removed comment to avoid misleading knowledge.

The solution that seemed to work was simply adding the __/nodeReuse:false__ switch to command line, but it's not the final solution.

Today after some other tests I think I've found the correct solution.

**Background**
Invoking msbuild with start-process seems to hang in some manner when msbuild process exits and control is left to the invoking console. I manually tested invoking msbuild from powershell in several ways, entirely bypassing Invoke-MsBuild script, and I came to the same misbehavior.

Then I tried to simply call start-process with the PassThru switch; in this way all performed tests seemed to work correctly, so I think that it's a problem related to invoking msbuild with the -Wait switch.

**The quick-and-dirty solution**
Call msbuild with the -PassThru switch, wait for the process to terminate, examine the results and you're done

**The real World**
When you call start-process with the -PassThru switch, process exit code could be unavailable. You have to play dirty with powershell but there is a bug that prevents you to obtain the exit code of a process with -PassThru switch. I found [a solution on stackoverflow](http://stackoverflow.com/questions/10262231/obtaining-exitcode-using-start-process-and-waitforexit-instead-of-wait) that involves storing process handle, so that you can obtain exit code.

There could be still an issue when storing process handle, as process could have been terminated before you store its handle; so I had to test if exit code is not null, and in that case I assume that process has correctly terminated.

**Conclusion**
With my modified script I can call Invoke-MsBuild in this way

`$buildResult = Invoke-MsBuild -Path $Project -MsBuildParameters $buildParameters -Use32BitMsBuild -PassThru -ShowBuildWindow`

and I can check results as I was calling it without -PassThru switch:

`if ($buildResult.BuildSucceeded -eq $true) ...`

All seems to work now. Let me know your opinions
[Invoke-MsBuild.psm1.txt](https://github.com/deadlydog/Invoke-MsBuild/files/521400/Invoke-MsBuild.psm1.txt)

.

**Comment by Jerome2606** [comment]
I had same issues, and I always use -PassThru option

It gives me something like that:

```
$parameterMSBuild = "/property:Configuration=" + $BuildConfigurationKey
if (![string]::IsNullOrEmpty($PublishProfileKey)){
    $parameterMSBuild = "/p:DeployOnBuild=true /p:PublishProfile=" + $PublishProfileKey + " /p:Configuration=" + $ConfigTransformConfigurationKey
}

$msbuildCommand = "/target:Clean;Build " + $parameterMSBuild

$buildProcess = Invoke-MsBuild -Path $WebAppFullPath -Params $msbuildCommand -PassThru 
# Set a timeout to 5 minutes (60 * 5)
$maximumRuntimeSeconds = 300
try
{
    $buildProcess | Wait-Process -Timeout $maximumRuntimeSeconds -ErrorAction Stop
    Write-Host "Process successfully completed within timeout."
}
catch
{
    Write-Warning -Message 'Process exceeded timeout, will be killed now.'
    $buildProcess | Stop-Process -Force
}

# Default buildLogDirectoryPath from module Invoke-MSBuild
$BuildLogDirectoryPath = $env:Temp

# Always get the full path to the Log files directory.
$BuildLogDirectoryPath = [System.IO.Path]::GetFullPath($BuildLogDirectoryPath)

$solutionFileName = (Get-ItemProperty -Path $WebAppFullPath).Name
$buildLogFilePath = (Join-Path -Path $BuildLogDirectoryPath -ChildPath $solutionFileName) + ".msbuild.log"
# Get if the build failed or not by looking at the log file.
$buildSucceeded = (((Select-String -Path $buildLogFilePath -Pattern "Build FAILED." -SimpleMatch) -eq $null) -and $buildProcess.ExitCode -eq 0)

if ($buildSucceeded -eq $true)
{ 
    Write-Host "Build completed successfully." 
}
elseif ($buildSucceeded -eq $false)
{ 
    Write-Host "Build failed. Check the build log file $buildLogFilePath for errors." 

    Write-Host "------------------------------------------------------------------------------------------------"
    Write-Host "Ending " $MyInvocation.MyCommand.Name

    Exit
}
elseif ($buildSucceeded -eq $null)
{ 
    Write-Host "Unsure if build passed or failed: $($buildResult.Message)" 
}
```

**Comment by deadlydog** [comment]
Sorry it took so long for me to see this. Apparently I didn't have my notifications configured properly in GitHub :(

Thanks for the detailed analysis and potential solutions/workarounds @StefanoBettega and @Jerome2606.

@StefanoBettega I've created a new `PotentialFixForMsBuildHangsWhenUsingPassThru` branch that includes the changes you added to the txt file you linked to. My concern is that it essentially changes the functionality of the PassThru switch. Instead of returning back the process that is executing the build, it instead waits for the process to exit, which is the same functionality as when not using the PassThru switch. Or have I misinterpreted this?

**Comment by timker** [comment]
we implemented something similar ( oops $process.HasExited with -PassThru)

We would have preferred be able to get a completed object back, but having it not hang was significantly better.

**Comment by StefanoBettega** [comment]
@deadlydog thanks for your answer.
Yes you got it: my suggested patch changes the way the PassThru switch is used, in fact the script will run until MSBuild process runs, and exits after the process is terminated. I changed the behavior because it was easier for me to use that switch to keep all the logging features you inserted, but maybe the right thing to to is modify the script execution mode to invoke MSBuild as it was used -PassThru switch, check the process-object returned from inside the script, and then behave exactly you did.

I was not able to determinate the real cause of this problem: I think is due to the fact that the PowerShell script is creating a console, that in turn spawn MSBuild and then all of the consoles which do the compilation work. This spanned-tree generates a lot of string output that is to be catched and parsed from PowerShell engine, causing it to delay its exit. If you look at its process stack using ProcessExplorer, you see that the root PowerShell process has a lot of threads running, doing something on console output (sorry, don't remember the exact stack function name). This could be the reason of the timeout: PowerShell has to wait for all the threads to finish their processing.

My patch is only a way to be sure that compilation works; we were planning to use it to produce NuGet packages of our libraries for internal use, so we had to be sure that all of the compilations were really executed.

Googling around a little bit more led me to find some issues with PowerShell and execution of batch files, expecially when the script (or launched process)has to deal with STDIN redirection. Maybe this is a starting point for further investigation...

Cheers,
Stefano

**Comment by iswix-llc** [comment]
I have used this module for dozens of builds without issue until day. I added a .sln to a build and msbuild reports time elapsed was 4.79 seconds but invoke-msbuild sits for about 15 minutes or so before continuing.  I tried the code from the PotentialFixForMsBuildHangsWhenUsingPassThru  branch and it hangs too.   Do I need to change my invocation when using this code?
It's very reproducible on this .sln file if you need me to report any debugging data.

**Comment by StefanoBettega** [comment]
Hello, I'm on vacancy now.
If I well remember, you have to invoke the script giving the passthrough option, even though it won't run in passthrough (asynchronous) mode but in synchronous mode.
Cheers,
Stefano

Il 9 agosto 2017 18:30:24 CEST, Christopher Painter <notifications@github.com> ha scritto:
>I have used this module for dozens of builds without issue until day. I
>added a .sln to a build and msbuild reports time elapsed was 4.79
>seconds but invoke-msbuild sits for about 15 minutes or so before
>continuing.  I tried the code from the
>PotentialFixForMsBuildHangsWhenUsingPassThru  branch and it hangs too. 
> Do I need to change my invocation when using this code?
>It's very reproducible on this .sln file if you need me to report any
>debugging data.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/deadlydog/Invoke-MsBuild/issues/11#issuecomment-321309046

-- 
Inviato dal mio dispositivo Android con K-9 Mail. Perdonate la brevità.

**Comment by iswix-llc** [comment]
I was able to grab this script and add -PassThru and solve the hang issue.  However two new issues were introduced that stumped me.  It wouldn't allow me to pass the params that specified to keep the logs directory and where to put them.  I kept getting powershell parameter binding issues.  I ended up hardcoding those values in the script and everything works now.  Wierd.  Enjoy your vacation!

**Comment by deadlydog** [comment]
Haven't had time to read through this post yet, but dumping it here as it may offer solutions to solve this problem as well. https://nftb.saturdaymp.com/fixing-msbuild-not-exiting/

**Comment by vip32** [comment]
i had the issue too, adding the UseSharedConfiguration=false property to the commandline fixed the issue.

**Comment by guillemsola** [comment]
Related issue in MS msbuild

https://github.com/Microsoft/msbuild/issues/2269

**Comment by peans99** [comment]
for me it was cause by the reuse of msbuild process I added /nr:false

-MsBuildParameters "$target /property:BuildInParallel=true /maxcpucount /nr:false"

**Comment by deadlydog** [comment]
Hey guys, I just pushed a potential fix out in version 2.6.4. Can you guys give it a try and let me know if you're still encountering the issue please. Thanks!

**Comment by jwi345** [comment]
Using version 2.6.2 I experienced a delay of approx. 10 minutes after MSBuild finished. Neither adding nr:/false or adding any other of the proposed options solved the issue. 

After reading this thread I updated to version 2.6.4. Now the same builds run in less than 1 minute so I assume that the fix solved the exit problem here. 

Thank you for working on it.

**Comment by deadlydog** [comment]
Awesome, thanks for confirming the issue is fixed! 👍
